### PR TITLE
Fix for set `DnsNameLabelReusePolicy`/`autoGeneratedDomainNameLabelScope` during creation of a new `ContainerGroup`.

### PR DESCRIPTION
**Pull request by v-hongli1**
- https://github.com/Azure/azure-sdk-for-java/issues/42341

# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java [hunk 1]
```diff
@@ -1179,7 +1182,18 @@ interface WithDnsPrefix {
              * @param dnsPrefix the DNS prefix to be used to create the FQDN for the container group
              * @return the next stage of the definition
              */
-            WithCreate withDnsPrefix(String dnsPrefix);
+            WithDnsNameLabelReusePolicy withDnsPrefix(String dnsPrefix);
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java:1185]: Why change the return value to `WithDnsNameLabelReusePolicy`? Is it must set?

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/CHANGELOG.md [hunk 2]
```diff
@@ -4,11 +4,7 @@
 
 ### Features Added
 
-### Breaking Changes
-
-### Bugs Fixed
-
-### Other Changes
+- Supported `autoGeneratedDomainNameLabelScop` method for `ContainerGroup`.
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-containerinstance/CHANGELOG.md:7]: check again!

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/pom.xml [hunk 3]
```diff
@@ -48,6 +48,7 @@
       --add-opens com.azure.core/com.azure.core.implementation.util=ALL-UNNAMED
     </javaModulesSurefireArgLine>
     <doclintMissingInclusion>-</doclintMissingInclusion>
+    <revapi.skip>true</revapi.skip>
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-containerinstance/pom.xml:51]: why?

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java [hunk 4]
```diff
@@ -1179,7 +1182,18 @@ interface WithDnsPrefix {
...
+
+        /** The stage of the container group definition allowing to specify the DNS name label reuse policy. */
+        interface WithDnsNameLabelReusePolicy {
+            /**
+             * Specifies the reuse policy of the DNS name label be to the object's domain name label can be reused in specified scope.
+             *
+             * @param reusePolicy the DNS name label reuse policy
+             * @return the next stage of the definition
+             */
+            WithCreate autoGeneratedDomainNameLabelScope(DnsNameLabelReusePolicy reusePolicy);
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java:1196]: prefix `with`

File: sdk/resourcemanager/azure-resourcemanager-samples/src/main/java/com/azure/resourcemanager/containerinstance/samples/ManageContainerInstanceWithAzureFileShareMount.java [hunk 5]
```diff
@@ -60,6 +61,7 @@ public static boolean runSample(AzureResourceManager azureResourceManager) {
                 .withVolumeMountSetting(volumeMountName, "/aci/logs/")
                 .attach()
                 .withDnsPrefix(aciName)
+                .autoGeneratedDomainNameLabelScope(DnsNameLabelReusePolicy.NOREUSE)
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-samples/src/main/java/com/azure/resourcemanager/containerinstance/samples/ManageContainerInstanceWithAzureFileShareMount.java:64]: there should be no need to update the samples.

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java (hunk 1 above)
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java:1185]: I tried to set DnsNameLabelReusePolicy alone, but it didn't work. It must be used together with Dns name label. Therefore, AutoGeneratedDomainNameLabelScop can only appear in the create step as a subsidiary attribute of DnsPrefix.

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/pom.xml (hunk 3 above)
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-containerinstance/pom.xml:51]: Changed the return value of `withDnsPrefix` to `WithDnsNameLabelReusePolicy` is a breaking change. I tried adding ignore items in `revapi.json` but it didn't seem to work. So I had to add `<revapi.skip>true</revapi.skip>` in `pom.xml`.

File: sdk/resourcemanager/azure-resourcemanager-samples/src/main/java/com/azure/resourcemanager/containerinstance/samples/ManageContainerInstanceWithAzureFileShareMount.java (hunk 5 above)
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-samples/src/main/java/com/azure/resourcemanager/containerinstance/samples/ManageContainerInstanceWithAzureFileShareMount.java:64]: https://github.com/Azure/azure-sdk-for-java/pull/42795#discussion_r1830325855

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java (hunk 1 above)
Comment (weidongxu-microsoft) [reply sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java:1185]: The change here means, if user sets DnsPrefix, they had to set AutoGeneratedDomainNameLabelScope. Are you sure this is correct?

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/pom.xml (hunk 3 above)
Comment (weidongxu-microsoft) [reply sdk/resourcemanager/azure-resourcemanager-containerinstance/pom.xml:51]: Nothing should break in this kind of feature PR.

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java (hunk 1 above)
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java:1185]: Yes, if reuse is not required, you can set it to `DnsNameLabelReusePolicy.NOREUSE`.

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java [hunk 6]
```diff
@@ -84,6 +84,9 @@ public interface ContainerGroup extends GroupableResource<ContainerInstanceManag
     /** @return the DNS configuration for the container group */
     DnsConfiguration dnsConfig();
 
+    /** @return the reuse policy of DSN name label */
+    DnsNameLabelReusePolicy dnsNameLabelReusePolicy();
```
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java:88]: should conform to the setter method's name

File: sdk/resourcemanager/azure-resourcemanager-samples/src/main/java/com/azure/resourcemanager/containerinstance/samples/ManageContainerInstanceWithAzureFileShareMount.java (hunk 5 above)
Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-samples/src/main/java/com/azure/resourcemanager/containerinstance/samples/ManageContainerInstanceWithAzureFileShareMount.java:64]: > [#42795 (comment)](https://github.com/Azure/azure-sdk-for-java/pull/42795#discussion_r1830325855)

Comment seems not related. 
We can set `DnsPrefix` without `autoGeneratedDomainNameLabelScope` right? By default, it's `noreuse` if not set.https://github.com/Azure/azure-sdk-for-java/blob/16bf2b788d4d2e1548497a99c50829a65664f462/sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/DnsNameLabelReusePolicy.java#L11
Could you help add a test case to validate this behavior?

Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-samples/src/main/java/com/azure/resourcemanager/containerinstance/samples/ManageContainerInstanceWithAzureFileShareMount.java:64]: @XiaofeiCao 
You mean to add a validation test case for a blank value, right?

Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-samples/src/main/java/com/azure/resourcemanager/containerinstance/samples/ManageContainerInstanceWithAzureFileShareMount.java:64]: Yeah. From service specs, if we don't specify a value for `autoGeneratedDomainNameLabelScope`, it'll be `Unsecure`. Please help validate.

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java (hunk 6 above)
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java]: Fixed in the new version.

File: sdk/resourcemanager/azure-resourcemanager-samples/src/main/java/com/azure/resourcemanager/containerinstance/samples/ManageContainerInstanceWithAzureFileShareMount.java (hunk 5 above)
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-samples/src/main/java/com/azure/resourcemanager/containerinstance/samples/ManageContainerInstanceWithAzureFileShareMount.java:64]: Fixed in the new version.

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java (hunk 4 above)
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java]: Fixed in the new version.

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/pom.xml (hunk 3 above)
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-containerinstance/pom.xml]: Fixed in the new version.

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/CHANGELOG.md (hunk 2 above)
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-containerinstance/CHANGELOG.md]: Fixed in the new version.

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java (hunk 1 above)
Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java:1185]: Please revert this breaking change. `autoGeneratedDomainNameLabelScope` is not required.

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/test/java/com/azure/resourcemanager/containerinstance/ContainerGroupTest.java [hunk 7]
```diff
@@ -225,4 +228,135 @@ public void testBeginCreateWithFileShareVolume() {
...
+            .withDnsPrefix(dnsNameLbl)
+            .withAutoGeneratedDomainNameLabelScope(DnsNameLabelReusePolicy.NOREUSE)
+            .create();
+
+        Assertions
+            .assertTrue(DnsNameLabelReusePolicy.NOREUSE.equals(containerGroup.autoGeneratedDomainNameLabelScope()));
+    }
+
+    @Test
+    public void testCreateWithDnsNameLabelUnSecure() {
```
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-containerinstance/src/test/java/com/azure/resourcemanager/containerinstance/ContainerGroupTest.java:257]: What is this test for?

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/test/java/com/azure/resourcemanager/containerinstance/ContainerGroupTest.java [hunk 8]
```diff
@@ -225,4 +228,135 @@ public void testBeginCreateWithFileShareVolume() {
         ContainerGroup containerGroup = acceptedContainerGroup.getSyncPoller().getFinalResult();
         Assertions.assertEquals(1, containerGroup.volumes().size());
     }
+
+    @Test
+    public void testCreateWithDnsNameLabelNoReuse() {
```
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-containerinstance/src/test/java/com/azure/resourcemanager/containerinstance/ContainerGroupTest.java:233]: Same with https://github.com/Azure/azure-sdk-for-java/pull/42795/files#r1830426052

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/test/java/com/azure/resourcemanager/containerinstance/ContainerGroupTest.java [hunk 9]
```diff
@@ -225,4 +228,135 @@ public void testBeginCreateWithFileShareVolume() {
...
+        ContainerGroup containerGroup = containerInstanceManager.containerGroups()
+            .define(containerGroupName)
+            .withRegion(region)
+            .withNewResourceGroup(rgName)
+            .withLinux()
+            .withPublicImageRegistryOnly()
+            .withoutVolume()
+            .withContainerInstance("nginx", 80)
+            .withDnsPrefix(dnsNameLbl)
+            .withAutoGeneratedDomainNameLabelScope(null)
```
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-containerinstance/src/test/java/com/azure/resourcemanager/containerinstance/ContainerGroupTest.java:356]: Remove this line. And combine this test with `testCreateWithDnsNameLabelScopeReuse`

File: sdk/resourcemanager/azure-resourcemanager-samples/src/main/java/com/azure/resourcemanager/containerinstance/samples/ManageContainerInstanceWithAzureFileShareMount.java (hunk 5 above)
Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-samples/src/main/java/com/azure/resourcemanager/containerinstance/samples/ManageContainerInstanceWithAzureFileShareMount.java:64]: Remove this line from samples.

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java [hunk 10]
```diff
@@ -1179,7 +1182,18 @@ interface WithDnsPrefix {
              * @param dnsPrefix the DNS prefix to be used to create the FQDN for the container group
              * @return the next stage of the definition
              */
-            WithCreate withDnsPrefix(String dnsPrefix);
+            WithDnsNameLabelReusePolicy withDnsPrefix(String dnsPrefix);
+        }
+
+        /** The stage of the container group definition allowing to specify the DNS name label reuse policy. */
+        interface WithDnsNameLabelReusePolicy {
```
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java:1189]: `WithAutoGeneratedDomainNameLabelScope`

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/test/java/com/azure/resourcemanager/containerinstance/ContainerGroupTest.java (hunk 8 above)
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-containerinstance/src/test/java/com/azure/resourcemanager/containerinstance/ContainerGroupTest.java]: Fixed in the [commit#6e748a3](https://github.com/Azure/azure-sdk-for-java/pull/42795/commits/6e748a33351eeda68205b4a03d69db268dd3d0ed)

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java (hunk 10 above)
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java]: Fixed in the [commit#6e748a3](https://github.com/Azure/azure-sdk-for-java/pull/42795/commits/6e748a33351eeda68205b4a03d69db268dd3d0ed)

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/test/java/com/azure/resourcemanager/containerinstance/ContainerGroupTest.java (hunk 9 above)
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-containerinstance/src/test/java/com/azure/resourcemanager/containerinstance/ContainerGroupTest.java]: Fixed in the [commit#6e748a3](https://github.com/Azure/azure-sdk-for-java/pull/42795/commits/6e748a33351eeda68205b4a03d69db268dd3d0ed)

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/test/java/com/azure/resourcemanager/containerinstance/ContainerGroupTest.java (hunk 7 above)
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-containerinstance/src/test/java/com/azure/resourcemanager/containerinstance/ContainerGroupTest.java]: Fixed in the [commit#6e748a3](https://github.com/Azure/azure-sdk-for-java/pull/42795/commits/6e748a33351eeda68205b4a03d69db268dd3d0ed)

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java [hunk 11]
```diff
@@ -1180,6 +1183,14 @@ interface WithDnsPrefix {
...
              */
             WithCreate withDnsPrefix(String dnsPrefix);
+
+            /**
+             * Specifies the reuse policy of the DNS name label be to the object's domain name label can be reused in specified scope.
+             *
+             * @param reusePolicy the DNS name label reuse policy
+             * @return the next stage of the definition
+             */
+            WithCreate withAutoGeneratedDomainNameLabelScope(DnsNameLabelReusePolicy reusePolicy);
```
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java:1193]: Can this property be updated?

Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java:1193]: The `ContainerGroup` interface does not contain any Update step, so I think `ContainerGroup` is a non-updatable service.
image

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/implementation/ContainerGroupImpl.java [hunk 12]
```diff
@@ -445,6 +446,14 @@ public ContainerGroupImpl withDnsPrefix(String dnsPrefix) {
         return this;
     }
 
+    @Override
+    public ContainerGroupImpl withAutoGeneratedDomainNameLabelScope(DnsNameLabelReusePolicy reusePolicy) {
+        if (reusePolicy != null) {
+            this.innerModel().ipAddress().withAutoGeneratedDomainNameLabelScope(reusePolicy);
+        }
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/implementation/ContainerGroupImpl.java:453]: make sure ipAddress is not null

https://github.com/Azure/azure-sdk-for-java/blob/6e748a33351eeda68205b4a03d69db268dd3d0ed/sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/implementation/ContainerGroupImpl.java#L441-L443

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/implementation/ContainerGroupImpl.java [hunk 13]
```diff
@@ -640,6 +649,15 @@ public DnsConfiguration dnsConfig() {
         return this.innerModel().dnsConfig();
     }
 
+    @Override
+    public DnsNameLabelReusePolicy autoGeneratedDomainNameLabelScope() {
+        if (this.innerModel().ipAddress() != null && this.innerModel().ipAddress().dnsNameLabel() != null) {
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/implementation/ContainerGroupImpl.java:654]: Just check `this.innerModel().ipAddress() != null`.

Give user what service gives SDK. No need to do extra check on "maybe related" property.

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java [hunk 14]
```diff
@@ -84,6 +84,9 @@ public interface ContainerGroup extends GroupableResource<ContainerInstanceManag
     /** @return the DNS configuration for the container group */
     DnsConfiguration dnsConfig();
 
+    /** @return the reuse policy of DSN name label */
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java:87]: javadoc need to write full description https://github.com/Azure/azure-sdk-for-java/issues/36596

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java]: Fixed in the new version.

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java (hunk 14 above)
Comment (weidongxu-microsoft) [reply sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java]: I mean, all javadoc has to have a "description" part.

```java
/**
 * <description>
 *
 * @return: ...
...
```

**Review by weidongxu-microsoft — APPROVED** [review]
LGTM

Please still wait for Xiaofei's approval. (he's OOF today, so you can work on other tasks first)

File: sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java (hunk 14 above)
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-containerinstance/src/main/java/com/azure/resourcemanager/containerinstance/models/ContainerGroup.java]: Fixed in the [commit#a9bf1bf](https://github.com/Azure/azure-sdk-for-java/pull/42795/commits/a9bf1bf61dfd459b0e3e909ae283257cb6b337b8)

**Review by XiaofeiCao — APPROVED** [review]
Please update PR description as well. There's no breaking change now.

**Pull request merged**